src/python/pants/backend/python/dependency_inference/rules.py

```diff
@@ -99,9 +109,28 @@ class InferPythonDependencies(InferDependenciesRequest):
     infer_from = PythonSources
 
 
+@dataclass(frozen=True)
+class InterpreterForPythonTargetRequest:
+    address: Address
+
+
+@rule
+async def interpreter_for_target(
+    request: InterpreterForPythonTargetRequest,
+    python_setup: PythonSetup,
+) -> PythonExecutable:
+    wrapped_tgt = await Get(WrappedTarget, Address, request.address)
+    compatibility = wrapped_tgt.target.get(PythonInterpreterCompatibility)
+    pex_constraints = PexInterpreterConstraints.create_from_compatibility_fields(
+        [compatibility], python_setup
+    )
+    return await Get(PythonExecutable, PexInterpreterConstraints, pex_constraints)
```

Comment:
While I haven't benchmarked, there's a non-trivial performance hit to running this rule. It will invoke a Pex subprocess that runs the equivalent of `pex --interpreter-constraint=blah`, i.e. with no requirements. We know that  Pex has some startup overhead. Even though the cache will be used for running this subprocess, this will still result in more overhead than we had before.

Instead, how about for Python 3 always using the Py38 AST, i.e. hardcoding it? This is what the above original comment was after, that the newer ASTs are backwards compatible with older Python versions. A Py38 AST should be able to parse Py35, Py36, and Py37 just fine. The main issue is that `async await` became keywords, but we already decided that's okay.

We want the target to be Py38, rather than say Py37, so that we ensure we never choke on Py38 code.

Reply:
I was under the impression that engine node memoization in pants will mean we only run this once per unique `constraints` setting, not once per target (or once per generated subtarget). Would that address your concern?

Reply:
> We want the target to be Py38, rather than say Py37, so that we ensure we never choke on Py38 code.
 
Could you perhaps elaborate on this? I would assume that if a user wanted to ensure their target was compatible with python 3.8, they would set the compatibility to that. The utility of getting the `PythonExecutable` is that we execute pex's interpreter selection on its constraints, just as would be the case in other contexts that the same target is used.

Reply:
Yes for running the subprocess, but you may still end up running more subprocesses than we intended. And you will still run the rule `interpreter_for_target` for each target, given that its inputs differ for each target.

> While I haven't benchmarked

Bump on this. I'm not blocking on this nor saying we must remove this rule, only that it could possibly cause issues, and it would be helpful to explore if we could mitigate or avoid adding more work for Pants to have to do. Dep inference is already fairly slow on the first run (mostly from creating the module mapping). But if we find we do need this logic, then we need it and can benchmark if necessary. 

What do you think about this proposal? 

> how about for Python 3 always using the Py38 AST,

Reply:
> I would assume that if a user wanted to ensure their target was compatible with python 3.8, they would set the compatibility to that. 

Our sole intent with this code is to read their import statements. The only reason we consider which Python version we parse with is so that the AST parser does not blow up when encountering syntax like `:=`. We don't actually care about respecting the constraint they want, unlike the MyPy implementation. We only need to extract out the imports, and to not blow up.

Reply:
> And you will still run the rule interpreter_for_target for each target, given that its inputs differ for each target.

`await Get(PythonExecutable, PexInterpreterConstraints, pex_constraints)` will still only be executed once per unique compatibility.

> how about for Python 3 always using the Py38 AST,

Ok. This makes me extremely uncomfortable, especially because further parser changes are [already happening in 3.9](https://docs.python.org/3.9/whatsnew/3.9.html#pep-617-new-parser) and [in 3.10](https://docs.python.org/3.10/whatsnew/3.10.html#pep604-new-type-operator).

But it would reduce the scope of this change a lot and avoid having to mix in `pex.interpreter.Python{Interpreter,Identity}`, which was annoying and gross, so I will make this change.

Reply:
Speaking of 3.9 being released today(!!), as soon as lib-cst is updated for 3.9, we will want to set the target to 3.9.

This works _so long_ as 3.9 continues to be able to parse 3.5-3.8 without issue. From what I know, that is true except for `async await` becoming reserved. Changes like `from __future__ import annotations` are fine; it's only syntax deprecations we're concerned about.